### DISCUSSION TITLE
Unable to create custom literal_processor for Date and DateTime

### DISCUSSION THREAD

**Discussion by villebro** [sqlalchemy/sqlalchemy#7002] [Usage Questions]
### Describe the bug

I'm trying to create a column type that can handle both `date` and `datetime` binds. While I'm able to extend the `DateTime` type to handle `datetime.datetime` and `Date` to handle `datetime.date` respectively, I'm unable to add support for `datetime.date` to the type that extends `DateTime` or `datetime.datetime` to the type that extends `Date`. Is this a bug, or am I doing something wrong?

### To Reproduce

```python
from datetime import date, datetime

import sqlalchemy as sa
from sqlalchemy.sql import column, desc, select, text
from sqlalchemy.types import Date, DateTime

engine = sa.create_engine('postgresql://uid:pwd@localhost')

def temporal_wrapper_factory(sqla_type):
    class TemporalWrapperType(sqla_type):
        def literal_processor(self, dialect):
            def process(value):
                if isinstance(value, (date, datetime)):
                    return f"'{str(value)}'"
                return super().process(value)

            return process

    return TemporalWrapperType

dt = date.today()
dttm = datetime.now()

dt_col = column('dt', type_=temporal_wrapper_factory(Date))
dttm_col = column('dttm', type_=temporal_wrapper_factory(DateTime))

test_cases = (
    ('DateTime with datetime', select(dttm_col).where(dttm_col == dttm)),
    ('DateTime with date', select(dttm_col).where(dttm_col == dt)),
    ('Date with datetime', select(dt_col).where(dt_col == dttm)),
    ('Date with date', select(dt_col).where(dt_col == dt)),
)

for descr, case in test_cases:
    print()
    print(f"Test case: {descr}")
    print(case)
    print()
    print("Compiled:")
    try:
        compiled = case.compile(engine, compile_kwargs={"literal_binds": True})
        print(compiled)
    except Exception as ex:
        print("FAIL!!!!")
        print(ex)
    print()
    print("-----")
```

### Error

Test case: DateTime with datetime
SELECT dttm 
WHERE dttm = :dttm_1

Compiled:
SELECT dttm 
WHERE dttm = '2021-09-08 14:42:44.317120'

-----

Test case: DateTime with date
SELECT dttm 
WHERE dttm = :dttm_1

Compiled:
FAIL!!!!
Don't know how to literal-quote value datetime.date(2021, 9, 8)

-----

Test case: Date with datetime
SELECT dt 
WHERE dt = :dt_1

Compiled:
FAIL!!!!
Don't know how to literal-quote value datetime.datetime(2021, 9, 8, 14, 42, 44, 317120)

-----

Test case: Date with date
SELECT dt 
WHERE dt = :dt_1

Compiled:
SELECT dt 
WHERE dt = '2021-09-08'

-----

### Versions

- OS: MacOS
- Python: 3.8.9
- SQLAlchemy: 1.4.23
- Database: None
- DBAPI: postgresql

**Comment by CaselIT** [comment]
Hi,

You need to provide a `literal_processor` only if you plan to literally compile the query.
The usual way would be to subclass TypeDecorator
https://docs.sqlalchemy.org/en/14/core/custom_types.html?highlight=type_descriptor#typedecorator-recipes

What was your use case?

**Reply by zzzeek** [reply]
I would need to see stack traces but I assume the problem above is that the dialect specific impl is being pulled in before it consults literal processor, so yes typedecorator should be used as it takes this contingency into account.  I wonder though if the documentation is still illustrating subclassing of the base types which really doesn't accomplish much these days we should make sure caveats like this are prominent

**Reply by CaselIT** [reply]
>  I wonder though if the documentation is still illustrating subclassing of the base types

From a quick look at that page, only when implementing custom operators: https://docs.sqlalchemy.org/en/14/core/custom_types.html?highlight=type_descriptor#redefining-and-creating-new-operators

**Reply by tboddyspargo** [reply]
FYI - I ran into this recently, too.

Example: 
```python
str(sa.func.cast('2022-01-01', sa.DATE).compile(dialect=postgresql.dialect, compile_kwargs=dict(literal_binds=True)))
```

My use case was just wanting to be able to log the "literal" query (as dialect-specific as possible) before executing it. It would be rather challenging to make sure that every dialect's Date/Datetime types were swapped out for TypeDecorators just to be able to print the compiled query with literal binds.

**Would it be reasonable for the generic (e.g. `sa.Date`) and/or sql standard/db-agnostic (`sa.DATE`) types to provide default literal bind processors to avoid hitting these kinds of issues?**

Otherwise, Is there a more "centralized" work-around (i.e. like a `@literal_bind_compiles(sa.Date)` option that could provide a fallback literal bind processor for a base/generic type)? Or is there a way to tell `ClauseElement.compile` to fall back to `StrCompileDialect` if the primary dialect's type can't compile the literal.

**EDIT:** FYI - another work-around is to wrap your string literal in `sa.literal`. I'm not entirely sure if that will have negative side-effects at execution time for different dialects, but at least the compilation works.
